File: [C]hashmap/[C]hashmap/383-ransom-note.py
class Solution:
    def canConstruct(self, ransomNote: str, magazine: str) -> bool:
        if len(ransomNote) > len(magazine):
            return False
        rchar_freq = [0 for _ in range(26)]
        mchar_freq = [0 for _ in range(26)]
        for c in ransomNote:
            rchar_freq[ord(c) - ord('a')] += 1
        for c in magazine:
            mchar_freq[ord(c) - ord('a')] += 1
        for i in range(26):
            if rchar_freq[i] > mchar_freq[i]:
                return False
        return True
    
# time O(n+m)
# space O(26) == O(1), constant
# using hashmap and store sth’s freq

File: [C]hashmap/[C]hashmap/383-ransom-note.java
class Solution {
    public boolean canConstruct(String ransomNote, String magazine) {
        HashMap<Character, Integer> rCharToFreq = new HashMap<>();
        HashMap<Character, Integer> mCharToFreq = new HashMap<>();
        for (int i = 0; i < ransomNote.length(); i++) {
            Character c = ransomNote.charAt(i);
            rCharToFreq.compute(c, (k, v) -> v == null ? 1 : v + 1);
        }
        for (int i = 0; i < magazine.length(); i++) {
            Character c = magazine.charAt(i);
            mCharToFreq.compute(c, (k, v) -> v == null ? 1 : v + 1);
        }
        for (Character c: rCharToFreq.keySet()) {
            if (mCharToFreq.containsKey(c) && rCharToFreq.get(c) <= mCharToFreq.get(c)) {
                continue;
            }
            return false;
        }
        return true;
    }
}

// time O(n+m)
// space O(26) == O(1), constant
// using hashmap and store sth’s freq